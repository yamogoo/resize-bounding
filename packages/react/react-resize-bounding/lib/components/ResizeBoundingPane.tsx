import {
  useMemo,
  useRef,
  useState,
  MutableRefObject,
  type PropsWithChildren,
  type CSSProperties,
  useEffect,
} from "react";

import {
  PaneDirections,
  PaneDirectionAliases,
  type PaneDirectionKey,
  type IResizeBoundingClassNames,
  type IStyles,
  type Options,
} from "../shared/typings";

import {
  paneBaseStyles,
  splitterBaseStyles,
} from "./ResizeBoundingPane.styles";

export interface PaneEmittedData {
  x: number;
  y: number;
  dir: string;
}

export interface Props extends PropsWithChildren {
  prefix: string;
  direction: PaneDirectionKey;
  options: Options;
  classNames: IResizeBoundingClassNames;
  styles?: Partial<IStyles>;
  dragStart: (data: PaneEmittedData) => void;
  dragMove: (data: PaneEmittedData) => void;
  dragEnd: (data: PaneEmittedData) => void;
  focus: (state: boolean) => void;
}

const checkIsHorizontal = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.LEFT} | ${PaneDirections.RIGHT} | ${PaneDirectionAliases.HORIZONTAL}]`,
  ).test(direction);

const checkIsVertical = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.TOP} | ${PaneDirections.BOTTOM} | ${PaneDirectionAliases.VERTICAL}]`,
  ).test(direction);

const ResizeBoundingPane = ({
  children,
  classNames,
  direction,
  options,
  dragStart,
  dragMove,
  dragEnd,
  focus,
}: Props) => {
  const refPane: MutableRefObject<HTMLDivElement | null> = useRef(null);

  const [isFocused, setIsFocused] = useState(false);
  const [isPressed, setIsPressed] = useState(false);

  let isResizing = false;

  const paneComputedStyle = useMemo(() => {
    const _width = options?.width ?? 1;

    if (refPane && _width) {
      const _areaWidth = options.activeAreaWidth ?? options.width;
      const _styles = paneBaseStyles(
        _width,
        _areaWidth,
        options?.position ?? "central",
      );
      const value: CSSProperties = _styles[direction as PaneDirections];
      return value;
    }

    return;
  }, [options.width, options.activeAreaWidth, options?.position, direction]);

  const splitterComputedStyle = useMemo(() => {
    const _width = options?.width;

    if (refPane && _width) {
      const _areaWidth = options.activeAreaWidth ?? options.width;

      const _styles = splitterBaseStyles(_width, _areaWidth);
      const value: CSSProperties = _styles[direction as PaneDirections];
      return value;
    }
  }, [options.width, options.activeAreaWidth, direction]);

  const containerComputedStyles = useMemo(() => {
    const isHorizontal = checkIsHorizontal(direction);
    return { transform: `rotate(${isHorizontal ? 90 : 0}deg)` };
  }, [direction]);

  const updateCursor = (state: boolean) => {
    let cursorActive: string;

    if (checkIsHorizontal(direction)) {
      cursorActive = options?.cursor?.horizontal ?? "auto";
    } else if (checkIsVertical(direction)) {
      cursorActive = options?.cursor?.vertical ?? "auto";
    } else cursorActive = "auto";

    if (refPane.current)
      refPane.current.style.cursor = state ? cursorActive : "auto";
  };

  /* * * Events * * */

  const emitFocus = (state: boolean) => {
    focus(state);
  };

  const onFocus = (e: PointerEvent, isFocused: boolean): void => {
    e.stopPropagation();
    onSelected(isFocused);
  };

  const onSelected = (state: boolean): void => {
    setIsFocused(state);
    updateCursor(state);
    emitFocus(state);
  };

  const onDragStart = (e: PointerEvent): void => {
    e.preventDefault();
    e.stopImmediatePropagation();

    if (!options.touchActions && e.pointerType === "touch") return;

    isResizing = true;
    setIsPressed(true);
    onSelected(isResizing);

    const el = e.currentTarget as HTMLDivElement;
    el.setPointerCapture(e.pointerId);

    dragStart({
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: direction,
    });

    const onDragMove = (e: PointerEvent): void => {
      dragMove({
        x: Math.round(e.clientX),
        y: Math.round(e.clientY),
        dir: direction,
      });
    };

    const onDragEnd = (e: PointerEvent): void => {
      isResizing = false;
      setIsPressed(false);

      const el = e.currentTarget as HTMLDivElement;
      el.releasePointerCapture(e.pointerId);

      el.removeEventListener("pointermove", onDragMove);
      el.removeEventListener("pointerup", onDragEnd);

      dragEnd({
        x: Math.round(e.clientX),
        y: Math.round(e.clientY),
        dir: direction,
      });
    };

    el.addEventListener("pointermove", onDragMove);
    el.addEventListener("pointerup", onDragEnd);
  };

  const onDragCancel = (e: PointerEvent): void => {
    const el = e.currentTarget as HTMLDivElement;

    isResizing = false;
    el.releasePointerCapture(e.pointerId);

    dragEnd({
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: direction,
    });
  };

  const addEventListeners = () => {
    const el = refPane.current;

    if (el) {
      el.addEventListener("pointerenter", (e) => onFocus(e, true));
      el.addEventListener("pointerleave", (e) => onFocus(e, false));

      el.addEventListener("pointerdown", onDragStart);
      el.addEventListener("pointercancel", onDragCancel);
    }
  };

  const removeEventListeners = () => {
    const el = refPane.current;

    if (el) {
      el.removeEventListener("pointerenter", (e) => onFocus(e, true));
      el.removeEventListener("pointerleave", (e) => onFocus(e, false));

      el.removeEventListener("pointerdown", onDragStart);
      el.addEventListener("pointercancel", onDragCancel);
    }
  };

  useEffect(() => {
    addEventListeners();

    return () => {
      removeEventListeners();
    };
  });

  return (
    <div
      ref={refPane}
      data-testid="resize-bounding-pane"
      className={`${classNames.pane}${isPressed || isFocused ? " active" : ""}${options.addStateClasses ? ` ${isPressed ? "pressed" : isFocused ? "focused" : "normal"}` : ""}`}
      style={paneComputedStyle}
    >
      <div
        data-testid="resize-bounding-splitter"
        className={classNames.splitter}
        style={splitterComputedStyle}
      >
        {(isFocused || !options?.knob?.normalHidden) && options?.knob?.show && (
          <div
            data-testid="resize-bounding-splitter-container"
            className={classNames.splitterContainer}
            style={containerComputedStyles}
          >
            <div data-testid="resize-bounding-knob" className={classNames.knob}>
              {children}
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default ResizeBoundingPane;
