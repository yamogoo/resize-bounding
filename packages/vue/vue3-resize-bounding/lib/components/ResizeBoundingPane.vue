<template>
  <div
    ref="refPane"
    data-testid="resize-bounding-pane"
    :class="[
      classNames.pane,
      { active: isFocused || isPressed },
      options.addStateClasses && {
        normal: !isPressed && !isFocused,
        focused: isFocused,
        pressed: isPressed,
      },
    ]"
    :style="[paneComputedStyle]"
  >
    <div
      data-testid="resize-bounding-splitter"
      :class="[classNames.splitter]"
      :style="[splitterComputedStyle]"
    >
      <div
        v-if="
          (isFocused || !options?.knob?.normalHidden) && options?.knob?.show
        "
        data-testid="resize-bounding-splitter-container"
        :class="[classNames.splitterContainer]"
        :style="[containerComputedStyles]"
      >
        <div data-testid="resize-bounding-knob" :class="[classNames.knob]">
          <slot v-if="$slots.default"></slot>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import {
  onMounted,
  onUnmounted,
  ref,
  type Ref,
  computed,
  type HTMLAttributes,
} from "vue";

import {
  SplitterPositions,
  type IStyles,
  type Options,
  type IResizeBoundingClassNames,
  PaneDirectionAliases,
} from "../shared/typings";

import {
  paneBaseStyles,
  splitterBaseStyles,
} from "./ResizeBoundingPane.styles";

const props = withDefaults(defineProps<Props>(), {
  direction: PaneDirections.RIGHT,
});

const emits = defineEmits<{
  (e: Emits.DRAG_START, data: PaneEmittedData): void;
  (e: Emits.DRAG_MOVE, data: PaneEmittedData): void;
  (e: Emits.DRAG_END, data: PaneEmittedData): void;
  (e: Emits.FOCUS, isFocused: boolean): void;
}>();

const refPane: Ref<HTMLDivElement | null> = ref(null);

const isFocused = ref(false),
  isPressed = ref(false);

let isResizing = false;

const paneComputedStyle = computed(() => {
  const _width = props.options?.width ?? 1;

  if (refPane.value && _width) {
    const _areaWidth = props.options.activeAreaWidth ?? props.options.width;
    const _styles = paneBaseStyles(
      _width,
      _areaWidth,
      props.options?.position ?? SplitterPositions.CENTER,
    );
    const value: HTMLAttributes["style"] =
      _styles[props.direction as PaneDirections];
    return value;
  }
  return;
});

const splitterComputedStyle = computed(() => {
  const _width = props.options?.width;

  if (refPane.value && _width) {
    const _areaWidth = props.options.activeAreaWidth ?? props.options.width;

    const _styles = splitterBaseStyles(_width, _areaWidth);
    const value: HTMLAttributes["style"] =
      _styles[props.direction as PaneDirections];
    return value;
  }
});

const containerComputedStyles = computed(() => {
  const isHorizontal = checkIsHorizontal(props.direction);
  return { transform: `rotate(${isHorizontal ? 90 : 0}deg)` };
});

const updateCursor = (state: boolean) => {
  let cursorActive: string;

  if (checkIsHorizontal(props.direction)) {
    cursorActive = props.options?.cursor?.horizontal ?? "auto";
  } else if (checkIsVertical(props.direction)) {
    cursorActive = props.options?.cursor?.vertical ?? "auto";
  } else cursorActive = "auto";

  if (refPane.value) refPane.value.style.cursor = state ? cursorActive : "auto";
};

const emitFocus = (state: boolean) => {
  emits(Emits.FOCUS, state);
};

const onFocus = (e: PointerEvent, isFocused: boolean): void => {
  e.stopPropagation();
  onSelected(isFocused);
};

const onSelected = (state: boolean): void => {
  isFocused.value = state;
  updateCursor(state);
  emitFocus(state);
};

const onDragStart = (e: PointerEvent): void => {
  e.preventDefault();
  e.stopImmediatePropagation();

  if (!props.options.touchActions && e.pointerType === "touch") return;

  isResizing = true;
  isPressed.value = true;

  onSelected(isResizing);

  const el = e.currentTarget as HTMLDivElement;
  el.setPointerCapture(e.pointerId);

  emits(Emits.DRAG_START, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });

  const onDragMove = (e: PointerEvent): void => {
    emits(Emits.DRAG_MOVE, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  const onDragEnd = (e: PointerEvent): void => {
    isResizing = false;
    isPressed.value = false;

    updateCursor(isResizing);

    const el = e.currentTarget as HTMLDivElement;
    el.releasePointerCapture(e.pointerId);

    el.removeEventListener("pointermove", onDragMove);
    el.removeEventListener("pointerup", onDragEnd);

    emits(Emits.DRAG_END, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  el.addEventListener("pointermove", onDragMove);
  el.addEventListener("pointerup", onDragEnd);
};

const onDragCancel = (e: PointerEvent): void => {
  const el = e.currentTarget as HTMLDivElement;

  isResizing = false;
  el.releasePointerCapture(e.pointerId);

  emits(Emits.DRAG_END, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });
};

const addEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.addEventListener("pointerenter", (e) => onFocus(e, true));
    el.addEventListener("pointerleave", (e) => onFocus(e, false));

    el.addEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

const removeEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.removeEventListener("pointerenter", (e) => onFocus(e, true));
    el.removeEventListener("pointerleave", (e) => onFocus(e, false));

    el.removeEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

onMounted(addEventListeners);
onUnmounted(removeEventListeners);

defineExpose({ refPane });
</script>

<script lang="ts">
import { PaneDirections, type PaneDirectionKey } from "../shared/typings";

export interface Props {
  prefix: string;
  direction: PaneDirectionKey;
  options: Options;
  classNames: IResizeBoundingClassNames;
  styles?: Partial<IStyles>;
}

export enum Emits {
  FOCUS = "focus",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}

export interface PaneEmittedData {
  x: number;
  y: number;
  dir: string;
}

const checkIsHorizontal = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.LEFT} | ${PaneDirections.RIGHT} | ${PaneDirectionAliases.HORIZONTAL}]`,
  ).test(direction);
const checkIsVertical = (direction: string): boolean =>
  new RegExp(
    `[${PaneDirections.TOP} | ${PaneDirections.BOTTOM} | ${PaneDirectionAliases.VERTICAL}]`,
  ).test(direction);
</script>
