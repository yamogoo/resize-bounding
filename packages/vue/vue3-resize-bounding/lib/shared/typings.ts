import fluentui from "@fluentui/merge-styles";

export interface Knob {
  show: boolean;
  normalHidden: boolean;
}

export enum SplitterPositions {
  CENTER = "central",
  EXTERNAL = "external",
  INTERNAL = "internal",
}

export type PanePosition = SplitterPositions | string;

export interface Options {
  prefix: string;
  width: number;
  activeAreaWidth?: number;
  position: PanePosition;
  cursor: Partial<
    Record<"horizontal" | "vertical", CSSStyleDeclaration["cursor"]>
  >;
  knob: Partial<Knob>;
  touchActions: boolean;
}

export type ResizeBoundingClassNames =
  | "container"
  | "pane"
  | "splitter"
  | "splitterContainer"
  | "knob";

export interface IResizeBoundingClassNames {
  container: string;
  pane: string;
  splitter: string;
  splitterContainer: string;
  knob: string;
}

export interface IStyles extends Record<ResizeBoundingClassNames, IStyle> {}
export type IStyle = fluentui.IStyle;
export type IStyleSet = fluentui.IStyleSet;

/* * * Pane * * */

export enum PaneDirectionAliases {
  HORIZONTAL = "h",
  VERTICAL = "v",
}

export enum PaneDirections {
  LEFT = "l",
  RIGHT = "r",
  BOTTOM = "b",
  TOP = "t",
}
