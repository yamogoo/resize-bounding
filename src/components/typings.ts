export interface BBResizeOptions {
  pane: {
    knob: {
      constantlyShow: boolean;
    };
  };
}

export interface BBResizeStyles {
  cursor: {
    active: {
      horizontal: string;
      vertical: string;
    };
  };
}
