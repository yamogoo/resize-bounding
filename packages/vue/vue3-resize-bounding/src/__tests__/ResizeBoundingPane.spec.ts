import { describe, test, expect, beforeEach, vi } from "vitest";
import { mount, shallowMount, VueWrapper } from "@vue/test-utils";
import { nextTick } from "vue";

import { DataTestIds, PREFIX } from "./setup";

import deepmerge from "deepmerge";

import ResizeBoundingPane, {
  Emits,
  type PaneEmittedData,
  type Props,
} from "../../lib/components/ResizeBoundingPane.vue";
import { PaneDirections } from "../../lib/shared/typings";

import {
  defaultOptions,
  getClassNames,
} from "../../lib/components/ResizeBounding.classNames";

import {
  paneBaseStyles,
  splitterBaseStyles,
} from "../../lib/components/ResizeBoundingPane.styles";

const classNames = getClassNames({});

const requiredProps: Props = {
  prefix: PREFIX,
  direction: PaneDirections.LEFT, //PaneDirectionAliases.HORIZONTAL,
  options: defaultOptions,
  classNames,
};

describe("ResizeBoundingPane", () => {
  beforeEach(() => {
    Element.prototype.setPointerCapture = vi.fn();
    Element.prototype.releasePointerCapture = vi.fn();
  });

  describe("classes", () => {
    describe("states", async () => {
      test(`should have ."active" class name`, async () => {
        const wrapper = shallowMount(ResizeBoundingPane, {
          props: requiredProps,
        });

        const paneEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);
        const stateInit = paneEl.classes("active");

        expect(stateInit).toBe(false);

        await paneEl.trigger("pointerenter");
        const state = paneEl.classes("active");

        expect(state).toBe(true);
        expect(state).toMatchSnapshot();
      });

      test('should not have ".normal" class name by default', async () => {
        const wrapper = shallowMount(ResizeBoundingPane, {
          props: requiredProps,
        });

        const splitterEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

        expect(splitterEl.classes("normal")).toBeFalsy();
        expect(splitterEl.classes("normal")).toMatchSnapshot();
      });

      const props: Props = deepmerge(requiredProps, {
        options: { addStateClasses: true },
      });

      const wrapper = shallowMount(ResizeBoundingPane, { props });

      test('should have ".normal" class name by default (if addStateClasses)', () => {
        const splitterEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

        expect(splitterEl.classes("normal")).toBeTruthy();
        expect(splitterEl.classes("normal")).toMatchSnapshot();
      });

      test('should have ".focused" class name (if addStateClasses)', async () => {
        const paneEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

        await paneEl.trigger("pointerenter");

        expect(paneEl.classes("focused")).toBeTruthy();
        expect(paneEl.classes("focused")).toMatchSnapshot();
      });

      test('should have ".pressed" class name (if addStateClasses)', async () => {
        const paneEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

        await paneEl.trigger("pointerdown");

        expect(paneEl.classes("pressed")).toBeTruthy();
        expect(paneEl.classes("pressed")).toMatchSnapshot();
      });
    });
  });

  describe("events", () => {
    describe("focus", async () => {
      const wrapper = shallowMount(ResizeBoundingPane, {
        props: deepmerge(requiredProps, { direction: PaneDirections.RIGHT }),
      });

      const rootEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

      test(`should emit "${Emits.FOCUS}" with state (true)`, async () => {
        await rootEl.trigger("pointerenter");
        const isFocused: boolean = (
          wrapper.emitted(Emits.FOCUS) as boolean[][]
        )[0][0];

        expect(isFocused).toBeTruthy();
        expect(isFocused).toMatchSnapshot();
      });

      test(`should emit "${Emits.FOCUS}" with state (false)`, async () => {
        await rootEl.trigger("pointerleave");
        const isFocused = (wrapper.emitted(Emits.FOCUS) as boolean[][])[1][0];

        expect(isFocused).toBeFalsy();
        expect(isFocused).toMatchSnapshot();
      });
    });

    describe("drag", () => {
      const wrapper = shallowMount(ResizeBoundingPane, {
        props: deepmerge(requiredProps, { direction: PaneDirections.RIGHT }),
      });

      const rootEl = wrapper.find(`[data-testid="${DataTestIds.SPLITTER}"]`);

      test(`should emit "${Emits.DRAG_START} -> ${Emits.DRAG_MOVE} -> ${Emits.DRAG_END}" with {x, y, dir}`, async () => {
        const checkData = (
          eventName: string,
          wrapper: VueWrapper,
          eventData: { clientX: number; clientY: number },
        ): PaneEmittedData => {
          const ev = wrapper.emitted(eventName) as PaneEmittedData[][];
          const data = ev[0][0];
          const { x, y } = data;

          const { clientX, clientY } = eventData;

          expect(data).toHaveProperty("x");
          expect(x).toBe(clientX);
          expect(data).toHaveProperty("y");
          expect(y).toBe(clientY);
          expect(data).toHaveProperty("dir");
          expect(data.dir).toBe(PaneDirections.RIGHT);
          return data;
        };

        const events: PaneEmittedData[] = [];

        const makeTrigger = async (
          event: string,
          eventData: { clientX: number; clientY: number },
          emittedEventName: string,
        ) => {
          await rootEl.trigger(event, eventData);
          events.push(checkData(emittedEventName, wrapper, eventData));
        };

        await rootEl.trigger("pointerenter", { clientX: 1, clientY: 1 });
        await makeTrigger(
          "pointerdown",
          { clientX: 1, clientY: 1 },
          Emits.DRAG_START,
        );
        await makeTrigger(
          "pointermove",
          { clientX: 2, clientY: 2 },
          Emits.DRAG_MOVE,
        );
        await makeTrigger(
          "pointerup",
          { clientX: 3, clientY: 3 },
          Emits.DRAG_END,
        );

        expect(events.length).toBe(3);
        expect(events).toMatchSnapshot();
      });
    });
  });

  describe("options", () => {
    describe("knob", () => {
      test("should always show knob", () => {
        const props: Props = deepmerge(requiredProps, {
          options: { knob: { show: true } },
        });

        const wrapper = shallowMount(ResizeBoundingPane, { props });

        const knobEl = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`);

        expect(knobEl.exists()).toBeTruthy();
        expect(knobEl.attributes("data-testid")).toMatchSnapshot();
      });

      test("should not render the knob if { normalHidden: true }", () => {
        const wrapper = shallowMount(ResizeBoundingPane, {
          props: deepmerge(requiredProps, {
            options: { knob: { show: true, normalHidden: true } },
          }),
        });

        const knobEl = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`);
        const isKnobExists = knobEl.exists();

        expect(isKnobExists).toBeFalsy();
        expect(isKnobExists).toMatchSnapshot();
      });

      test.each([
        [PaneDirections.LEFT, 90],
        [PaneDirections.RIGHT, 90],
        [PaneDirections.TOP, 0],
        [PaneDirections.BOTTOM, 0],
      ])("pane (--%s) should rotated to (%s deg)", (direction, deg) => {
        const props: Props = deepmerge(requiredProps, {
          direction,
          options: { knob: { show: true } },
        });

        const wrapper = shallowMount(ResizeBoundingPane, { props });

        const knobEl = wrapper.find(
          `[data-testid="${DataTestIds.SPLITTER_CONTAINER}"]`,
        );
        const styles = knobEl.attributes("style");

        expect(styles).toContain(`(${deg}deg)`);
        expect(styles).toMatchSnapshot();
      });
    });
  });

  describe("styles", () => {
    describe("cursor", () => {
      test('default cursor style should be "auto"', async () => {
        const props: Props = requiredProps;

        const wrapper = shallowMount(ResizeBoundingPane, { props });
        await wrapper.setProps({ direction: PaneDirections.RIGHT });

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

        await rootEl.trigger("pointerenter");
        await rootEl.trigger("pointerleave");

        const cursorStyle = rootEl.attributes("style");

        expect(cursorStyle).toContain("auto");
        expect(cursorStyle).toMatchSnapshot();
      });

      const checkCursorActive = async (
        wrapper: VueWrapper,
        props: Partial<Props>,
        comparedCursorStyle: string,
      ): Promise<void> => {
        await wrapper.setProps(props);

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);
        await rootEl.trigger("pointerenter");

        const cursorStyle = rootEl.attributes("style");

        expect(cursorStyle).toContain(comparedCursorStyle);
        expect(cursorStyle).toMatchSnapshot();
      };

      describe("active", () => {
        describe("default styles", () => {
          test.each([PaneDirections.RIGHT, PaneDirections.LEFT])(
            'default cursor style should be "col-resize" (--%s)',
            async (direction) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: requiredProps,
              });

              checkCursorActive(wrapper, { direction }, "col-resize");
            },
          );

          test.each([PaneDirections.TOP, PaneDirections.BOTTOM])(
            'default cursor style should be "row-resize" (--%s)',
            async (direction) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: requiredProps,
              });

              checkCursorActive(wrapper, { direction }, "row-resize");
            },
          );
        });

        describe("custom styles (options)", () => {
          const CUSTOM_ACTIVE_CURSOR = "ew-resize";

          test.each([...Object.values(PaneDirections)])(
            `custom cursor style should be "${CUSTOM_ACTIVE_CURSOR}" (--%s)`,
            async (direction) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: requiredProps,
              });

              checkCursorActive(
                wrapper,
                {
                  direction,
                  options: {
                    cursor: {
                      vertical: CUSTOM_ACTIVE_CURSOR,
                      horizontal: CUSTOM_ACTIVE_CURSOR,
                    },
                    width: 0,
                    position: "central",
                    knob: {},
                    prefix: "",
                    touchActions: false,
                  },
                },
                CUSTOM_ACTIVE_CURSOR,
              );
            },
          );

          const SIZE = 12;
          const paneMap = paneBaseStyles(SIZE, SIZE, "central");
          const splitterMap = splitterBaseStyles(SIZE, SIZE);

          test.each([
            [PaneDirections.LEFT, paneMap.l],
            [PaneDirections.RIGHT, paneMap.r],
            [PaneDirections.TOP, paneMap.t],
            [PaneDirections.BOTTOM, paneMap.b],
          ])(
            `the pane (--%s) should have "%s" styles`,
            async (direction, styles) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: deepmerge(requiredProps, {
                  direction,
                  options: { width: SIZE },
                }),
              });

              await nextTick();

              const paneEl = wrapper.find(
                `[data-testid="${DataTestIds.PANE}"]`,
              );
              const style = paneEl.attributes("style");

              if (styles) {
                for (const [k, v] of Object.entries(styles)) {
                  expect(style).toContain(`${k}: ${v}`);
                }

                expect(style).toMatchSnapshot();
              }
            },
          );

          test.each([
            [PaneDirections.LEFT, splitterMap.l],
            [PaneDirections.RIGHT, splitterMap.r],
            [PaneDirections.TOP, splitterMap.t],
            [PaneDirections.BOTTOM, splitterMap.b],
          ])(
            `the pane splitter (--%s) should have "%s" styles`,
            async (direction, styles) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: deepmerge(requiredProps, {
                  direction,
                  options: { width: SIZE },
                }),
              });

              await nextTick();

              const splitterEl = wrapper.find(
                `[data-testid="${DataTestIds.SPLITTER}"]`,
              );
              const style = splitterEl.attributes("style");

              if (styles) {
                for (const [k, v] of Object.entries(styles)) {
                  expect(style).toContain(`${k}: ${v}`);
                }

                expect(style).toMatchSnapshot();
              }
            },
          );
        });
      });
    });
  });

  describe("slots", () => {
    test.each(["<p>knob</p>"])("should render knob slot (%s)>", (knobSlot) => {
      const wrapper = mount(ResizeBoundingPane, {
        props: deepmerge(requiredProps, {
          direction: "r",
          options: {
            knob: {
              show: true,
            },
          },
        }),
        slots: {
          default: knobSlot,
        },
      });

      const text = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`).text();
      expect(text).toBe("knob");
      expect(text).toMatchSnapshot();
    });
  });
});
