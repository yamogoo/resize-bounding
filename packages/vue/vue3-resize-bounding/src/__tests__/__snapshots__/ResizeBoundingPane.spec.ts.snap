// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResizeBoundingPane > classes > states > should have ".focused" class name (if addStateClasses) 1`] = `true`;

exports[`ResizeBoundingPane > classes > states > should have ".normal" class name by default (if addStateClasses) 1`] = `true`;

exports[`ResizeBoundingPane > classes > states > should have ".pressed" class name (if addStateClasses) 1`] = `true`;

exports[`ResizeBoundingPane > classes > states > should have ."active" class name 1`] = `true`;

exports[`ResizeBoundingPane > classes > states > should not have ".normal" class name by default 1`] = `false`;

exports[`ResizeBoundingPane > events > drag > should emit "drag:start -> drag:move -> drag:end" with {x, y, dir} 1`] = `
[
  {
    "dir": "r",
    "x": 1,
    "y": 1,
  },
  {
    "dir": "r",
    "x": 2,
    "y": 2,
  },
  {
    "dir": "r",
    "x": 3,
    "y": 3,
  },
]
`;

exports[`ResizeBoundingPane > events > focus > should emit "focus" with state (false) 1`] = `false`;

exports[`ResizeBoundingPane > events > focus > should emit "focus" with state (true) 1`] = `true`;

exports[`ResizeBoundingPane > options > knob > pane (--b) should rotated to (0 deg) 1`] = `"transform: rotate(0deg);"`;

exports[`ResizeBoundingPane > options > knob > pane (--l) should rotated to (90 deg) 1`] = `"transform: rotate(90deg);"`;

exports[`ResizeBoundingPane > options > knob > pane (--r) should rotated to (90 deg) 1`] = `"transform: rotate(90deg);"`;

exports[`ResizeBoundingPane > options > knob > pane (--t) should rotated to (0 deg) 1`] = `"transform: rotate(0deg);"`;

exports[`ResizeBoundingPane > options > knob > should always show knob 1`] = `"resize-bounding-knob"`;

exports[`ResizeBoundingPane > options > knob > should not render the knob if { normalHidden: true } 1`] = `false`;

exports[`ResizeBoundingPane > slots > should render knob slot (<p>knob</p>)> 1`] = `"knob"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--b) 1`] = `"cursor: ew-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--l) 1`] = `"cursor: ew-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--r) 1`] = `"cursor: ew-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--t) 1`] = `"cursor: ew-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane (--b) should have "{ left: '0px', bottom: '-6px', width: '100%', height: '12px' }" styles 1`] = `"left: 0px; bottom: -6px; width: 100%; height: 12px;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane (--l) should have "{ top: '0px', left: '-6px', width: '12px', height: '100%' }" styles 1`] = `"top: 0px; left: -6px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane (--r) should have "{ top: '0px', right: '-6px', width: '12px', height: '100%' }" styles 1`] = `"top: 0px; right: -6px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane (--t) should have "{ left: '0px', top: '-6px', width: '100%', height: '12px' }" styles 1`] = `"left: 0px; top: -6px; width: 100%; height: 12px;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane splitter (--b) should have "{ top: '0px', height: '12px', width: '100%' }" styles 1`] = `"top: 0px; height: 12px; width: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane splitter (--l) should have "{ right: '0px', width: '12px', height: '100%' }" styles 1`] = `"right: 0px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane splitter (--r) should have "{ left: '0px', width: '12px', height: '100%' }" styles 1`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > custom styles (options) > the pane splitter (--t) should have "{ bottom: '0px', height: '12px', width: '100%' }" styles 1`] = `"bottom: 0px; height: 12px; width: 100%;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "col-resize" (--l) 1`] = `"top: 0px; left: -2px; width: 4px; height: 100%; cursor: col-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "col-resize" (--r) 1`] = `"top: 0px; right: -2px; width: 4px; height: 100%; cursor: col-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "row-resize" (--b) 1`] = `"left: 0px; bottom: -2px; width: 100%; height: 4px; cursor: row-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > active > default styles > default cursor style should be "row-resize" (--t) 1`] = `"left: 0px; top: -2px; width: 100%; height: 4px; cursor: row-resize;"`;

exports[`ResizeBoundingPane > styles > cursor > default cursor style should be "auto" 1`] = `"top: 0px; right: -2px; width: 4px; height: 100%; cursor: auto;"`;
